File: frontend/src/lib/api/valuation.ts
import { protectedApiClient } from "./protected-client";

// Types
export interface SKUValuation {
  sku_code: string;
  name: string;
  total_qty: number;
  avg_cost: string;
  total_value: string;
  currency: string;
}

export interface SKUValuationResponse {
  items: SKUValuation[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface TotalValuation {
  total_value: number;
  currency: string;
  method: string;
  method_full_name: string;
  timestamp: string;
}

export interface COGSResponse {
  total_cogs: number;
  currency: string;
  timestamp: string;
  delta_percentage?: number;
  sku_code?: string;
  period_start?: string;
  period_end?: string;
}

export interface COGSResponse {
  total_cogs: number;
  currency: string;
  timestamp: string;
  delta_percentage?: number;
  sku_code?: string;
  period_start?: string;
  period_end?: string;
}

export interface SKUValuationParams {
  page?: number;
  size?: number;
}

export interface COGSParams {
  start_date?: string;
}

// API Functions
export async function getSKUValuations(
  params?: SKUValuationParams
): Promise<SKUValuationResponse> {
  const queryParams = new URLSearchParams();
  if (params?.page !== undefined) {
    queryParams.append('page', params.page.toString());
  }
  if (params?.size !== undefined) {
    queryParams.append('size', params.size.toString());
  }
  
  const url = `/valuation/skus${queryParams.toString() ? `?${queryParams}` : ''}`;
  return protectedApiClient<SKUValuationResponse>(url);
}

export async function getTotalValuation(): Promise<TotalValuation> {
  return protectedApiClient<TotalValuation>("/valuation");
}

export async function getCOGS(params?: COGSParams): Promise<COGSResponse> {
  const queryParams = new URLSearchParams();
  if (params?.start_date) {
    queryParams.append('start_date', params.start_date);
  }
  
  const url = `/valuation/cogs${queryParams.toString() ? `?${queryParams}` : ''}`;
  return protectedApiClient<COGSResponse>(url);
}

File: backend/app/schemas/valuation.py
from datetime import date, datetime
from typing import List, Optional
from pydantic import BaseModel, field_serializer

from  app.schemas.common import TrendPoint


class InventoryValuationRow(BaseModel):
    sku_code: str
    name: str
    total_qty: int
    avg_cost: float
    total_value: float
    currency: str
    

class ValuationHeader(BaseModel):
    total_value: float
    currency: str
    method: str
    method_full_name: str
    timestamp: datetime

    @field_serializer('timestamp')
    def serialize_dt(self, dt: datetime):
        return dt.isoformat()


class COGSResponse(BaseModel):
    total_cogs: float
    currency: str
    sku_code: Optional[str] = None
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    delta_percentage: float | None = None
    timestamp: str
    

class TrendPoint(BaseModel):
    date: date
    cogs: float

class COGSTrendResponse(BaseModel):
    oldest_data_point: Optional[date] = None
    points: List[TrendPoint]
    